### DISCUSSION TITLE
Mesh slot -1?

### DISCUSSION THREAD

**Discussion by Gil80** [Jyers/Marlin#1233] [Q&A] [answered]
Using UBL 10x10 firmware.
In mesh bed leveling I see Mesh slot -1 - see here: https://ibb.co/92CD87g

Why -1 and not 0? or 1?
Is this a bug?

EDIT: Seems that my mesh settings are not saved after I switch the printer off. I use BLTouch and at the end of the mesh leveling process I save the EPROM.
If I restart the printer, the tick box in Active Mesh Leveling is off and the Mesh Slot is set to -1.

**Comment by JoshRawdon** [comment]
I came here to post the same thing. I made a new 10x10 mesh, saved the mesh and printed something. Turned off the printer. Next day I insert the same SD card and go to print something else, and Leveling Active box is no longer ticked, and when I go to view mesh, it's all 0.00. I tried Load Mesh, and still nothing.

**Reply by gothcha** [reply]
Did you set the mesh slot to 0 first and then save the settings under the Control Menu? This will ensure mesh slot zero is selected when you power on next time and load the mesh. If you do not save the mesh setting it will remain set to -1 in the EEPROM and you will get the behaviour that you are reporting. Saving the mesh data itself after creating the mesh does not store the mesh number. 
Levelling Active will be set when you process the M420 S1 or G29 A in the start-up code, setting it in the menu I do not find is very beneficial unless maybe you need to do manual moves from the screen?

**Comment by Jyers** [comment, marked as answer]
This is just how marlin deals with leveling, On reboot, leveling is disabled. To make sure your mesh is saved, select load mesh and then check mesh viewer.

This is why we need M420/G29 L0 G29 A in the start gcode, as that makes sure leveling is enabled and the mesh is loaded before every print.

**Reply by Gil80** [reply]
G29 A is not sufficient? We need G29 L0 as well?
I thought that if I want to use the stored mesh I have in L1, I just use the LCD menu to do that and not the G code.

As a side note, I find UBL to be very buggy and that M420 S1 doesn't recall the bed leveling. G29 A works instead.
Also, I don't know if it's the firmware or the BLTouch but on three occasions I had the nozzle crashing into the bed when I started a new print job right after cancelling a previous job.

I think Bilinear is a safer option.

**Reply by Gil80** [reply]
@Jyers I don't recall that with the stock Merlin the mesh level is disabled after each reboot. It's always enabled. Despite that, it is granted that G29 A must be in the start g-code.
But the LCD menu, I always recall having the mesh level activated, after a reboot of course.

**Reply by Jyers** [reply]
@Gil80 In the marlin documentation they recommend adding L0, but you could get by without it, it's just good practice.

I can assure you that stock Marlin the mesh is disabled after reboot... Some implementations may force it to try to enable, but marlins default behavior is to have is disabled after a reboot.

**Reply by PapsDizzy** [reply]
so wich command i must set in cura to load the createt mesh from the E3V2-BLTouch-5x5-HS-v4.2.2-v1.3.5. firmware?
im confused, the wiki says M420 S1, here its told M420/G29 L0 G29 A.
so when i have created a mesh with the bl-touch from the level menü , wich commands must in the gcode in cura?
i only have the M420 S1 in the code, is this correct?

**Reply by adam3654** [reply]
> so wich command i must set in cura to load the createt mesh from the E3V2-BLTouch-5x5-HS-v4.2.2-v1.3.5. firmware?
> im confused, the wiki says M420 S1, here its told M420/G29 L0 G29 A.
> so when i have created a mesh with the bl-touch from the level menü , wich commands must in the gcode in cura?
> i only have the M420 S1 in the code, is this correct?

After G28, use 

G29 L0 ; load mesh from slot 0
G29 A ; activate mesh
G29 J ; tilt mesh

**Reply by PapsDizzy** [reply]
> 
> 
> > so wich command i must set in cura to load the createt mesh from the E3V2-BLTouch-5x5-HS-v4.2.2-v1.3.5. firmware?
> > im confused, the wiki says M420 S1, here its told M420/G29 L0 G29 A.
> > so when i have created a mesh with the bl-touch from the level menü , wich commands must in the gcode in cura?
> > i only have the M420 S1 in the code, is this correct?
> 
> After G28, use
> 
> G29 L0 ; load mesh from slot 0
> G29 A ; activate mesh
> G29 J ; tilt mesh
ok but the wiki tells me i have to use M420 S1 
see here [wiki](https://github.com/Jyers/Marlin/wiki/Configuration-Details#bl-touch)

**Reply by adam3654** [reply]
> ok but the wiki tells me i have to use M420 S1 
> see here [https://github.com/Jyers/Marlin/wiki/Configuration-Details#bl-touch](url)

They are equivalent as far as I know, and G29 always worked for me. Early on I had difficulty when using M420 S1.

**Reply by PapsDizzy** [reply]
> They are equivalent as far as I know, and G29 always worked for me. Early on I had difficulty when using M420 S1.

ok i will test it.

**Reply by PapsDizzy** [reply]
so its look like both works for me.

**Reply by Jyers** [reply]
G29 A etc is only for use with UBL configs. M420 S1 should be used with all others. This is detailed on the wiki

**Reply by Gil80** [reply]
> 
> 
> G29 A etc is only for use with UBL configs. M420 S1 should be used with all others. This is detailed on the wiki

I think it should be made more explicit and bold that G29 A is for UBL while M420 S1 is for the rest.

**Reply by Jyers** [reply]
@Gil80 It is clearly noted on the wiki you need `M420 S1` for BL Touch or Manual Mesh and `M420 S1 or G29 A` for UBL.

I suppose I could add an extra note somewhere to try to clarify it more, but it is definitely documented. Plus all this is just basic Marlin info which can be readily found online or all in one place on the [Marlin website](https://marlinfw.org/)

**Reply by PapsDizzy** [reply]
so one qusetion M420 S1 enables the mesh. can i also use M420 L1 S1 to load an enable? or is it stupid to load and activate?

**Reply by Jyers** [reply]
Unless you are using UBL there is no need to load the mesh. The reason you need to load the mesh in UBL is because you are able to store multiple meshes using that system. If you're using UBL you load the mesh (slot 0) with G29 L0, if you're not using UBL, simply use M420 S1

**Reply by Gil80** [reply]
> 
> 
> Unless you are using UBL there is no need to load the mesh. The reason you need to load the mesh in UBL is because you are able to store multiple meshes using that system. If you're using UBL you load the mesh (slot 0) with G29 L0, if you're not using UBL, simply use M420 S1

But is it true that for UBL I can use G29 A and in the LCD menu select whichever Mesh Slot I need based on the bed I'm using?

**Reply by Jyers** [reply]
@Gil80 Yes, that should work well
